C/tests/c4DatabaseTest.cc

```diff
@@ -947,19 +949,86 @@ TEST_CASE("Database Upgrade From 2.8", "[Database][Upgrade][C]") {
     C4DatabaseConfig2 config = {slice(TempDir()), withFlags};
-    C4Database *db;
+    c4::ref<C4Database> db;
     auto name = C4Test::copyFixtureDB(kVersionedFixturesSubDir + dbPath);
     C4Log("---- copy Fixture to: %s/%s", TempDir().c_str(), name.asString().c_str());
     C4Error err;
     db = c4db_openNamed(name, &config, WITH_ERROR(&err));
     CHECK(db);
 
-    Doc info(alloc_slice(c4db_getIndexesInfo(db, WITH_ERROR(&err))));
-    CHECK(info.asArray().count() == 1);
-    Dict index = info.asArray()[0].asDict();
-    REQUIRE(index);
-    CHECK(index["name"].asString() == "index1");
-    CHECK(index["type"].asInt() == kC4ValueIndex);
+    std::function<void(C4Database*)> checkIndex = [](C4Database* db){
+        C4Error err;
+        Doc info(alloc_slice(c4db_getIndexesInfo(db, WITH_ERROR(&err))));
+        CHECK(info.asArray().count() == 1);
+        Dict index = info.asArray()[0].asDict();
+        REQUIRE(index);
+        CHECK(index["name"].asString() == "index1");
+        CHECK(index["type"].asInt() == kC4ValueIndex);
+    };
+    checkIndex(db);
+
+    REQUIRE(c4db_close(db, WITH_ERROR()));
+
+    // Reopen the database and edit the first document
+
+    db = c4db_openNamed(name, &config, WITH_ERROR(&err));
+    CHECK(db);
+    CHECK(c4db_getDocumentCount(db) == 2);
+
+    // Get the first document
+
+    C4EnumeratorOptions options = kC4DefaultEnumeratorOptions;
+    c4::ref<C4DocEnumerator> e = c4db_enumerateAllDocs(db, &options, ERROR_INFO());
+    CHECK(c4enum_next(e, ERROR_INFO(&err)));
+
+    c4::ref<C4Document> doc1 = c4enum_getDocument(e, &err);
+    CHECK(doc1);
+    alloc_slice doc1ID = doc1->docID;
+    alloc_slice revID1 = doc1->revID;
+
+    // Get the revision body
+
+    Doc fl_docRev1(alloc_slice(c4doc_getRevisionBody(doc1)), kFLTrusted, c4db_getFLSharedKeys(db));
+    CHECK(fl_docRev1.root().type() == kFLDict);
+    Dict rev1Dict = fl_docRev1.root().asDict();
+
+    // Make a mutable copy of rev1BodyDict and edit it
+
+    MutableDict mdict2 = rev1Dict.mutableCopy();
+    vector<alloc_slice> keys;
+    for (MutableDict::iterator it = mdict2.begin(); it != mdict2.end(); ++it) {
+        slice key = it.keyString();
+        keys.push_back(alloc_slice(key));
+    }
+    // Update the values of mdict2
+    for (vector<alloc_slice>::const_iterator it = keys.begin(); it != keys.end(); ++it) {
+        Value val = mdict2.get(*it);
+        string valStr = val.toJSONString() + "-update";
+        mdict2.set(*it, valStr);
+    }
+
+    std::function<FLSliceResult(C4Database*, Dict&)> encodeDict =
+    [](C4Database* db, Dict& dict) {
+        TransactionHelper t(db);
+        FLEncoder enc = c4db_getSharedFleeceEncoder(db);
+        FLEncoder_WriteValue(enc, dict);
+        FLError flError = kFLNoError;
+        auto ret = FLEncoder_Finish(enc, &flError);
+        CHECK(flError == kFLNoError);
+        return ret;
+    };
+
+    // Encode the modified dict:
+    FLSliceResult fl_dict2 = encodeDict(db, mdict2);
+
```

Comment:
Please use the utility methods in C4Test to update the document. I don't see a need to write all this code just to test updating a document,